### PR TITLE
[WIP] hillst/noodles bionemo

### PR DESCRIPTION
**Pull request by skothenhill-nv**
Initial commit!

- adds the PyO3 wrapper around noodles-fasta
- adds a python class that mimics the dict-like behavior of pyfaidx
- adds a long test (best used with hg38 https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/), or use whatever big reference.

Build:
```
cd sub-packages/bionemo-noodles/noodles_fasta_wrapper
maturin develop
```
This should install it into your local python environment, then you can run the associated tests!

TODO:
- flesh out equality tests with pyfaidx
- profile performance
- get the build system working correctly

### CODE REVIEW

**Review by malcolmgreaves — APPROVED** [review]
🚀

File: sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            region_str = f"{self.seqid}:{start + 1}-{stop}"  # +1 for 1-based indexing
+            return self.reader.query_region(region_str)
+
+        elif isinstance(key, int):
+            # Handle single integer for single nucleotide queries
+            if key < 0 or key >= self.length:
+                raise IndexError(f"Position {key} is out of bounds for '{self.seqid}' with length {self.length}.")
+
+            # Query single nucleotide by creating a 1-length region
+            region_str = f"{self.seqid}:{key + 1}-{key + 1}"  # +1 for 1-based indexing
```
Comment (jstjohn) [inline sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py:29]: You have a test that covers this, right?  This kind of thing is so easy to get wrong. I'd like to see a test covering slices and single index lookups.

Comment (jstjohn) [reply sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py:29]: Seems right though, to be clear! I just don't trust myself with this kind of thing. Also someone in the future could easily break it and not realize.

File: sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
+from noodles_fasta_wrapper import IndexedFastaReader
+
+class SequenceAccessor:
+    # NOTE: we could totally handle this stuff in Rust if we want.
+    def __init__(self, reader, seqid, length):
+        self.reader = reader
+        self.seqid = seqid
+        self.length = length
+
+    def __getitem__(self, key):
```
Comment (jstjohn) [inline sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py:10]: We should be clear about how we're 0 based in docs. Maybe make a docstring for the class and spell that out.

File: sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        print(index['chr1'][1:10000])
+    except Exception:
+        pass
+
+    try:
+        print(index['chr1'][1])
+    except Exception:
+        pass
+
+def test_process_parallel_bug():
```
Comment (jstjohn) [inline sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py:71]: oh here are the tests! why not put them in the `tests` dir?

File: sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+    For a good solution we need three things:
+        1) Safe index creation, in multi-process or multi-node scenarios, this should be restricted to a single node where all workers block until it is complete (not implemented above)
+        2) Index object instantion must be fast.
+        3) Read-only use of the index object must be both thread safe and process safe with python.
+    '''
+    from multiprocessing import Pool
+
+    # NOTE: taken from github issue above, this reproduces the issue with inconsistent reads in parallel.
+    fasta = NvFaidx('hg38.fa')
```
Comment (jstjohn) [inline sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py:89]: You'll move to the `sample.fasta` right?

**Comment by jstjohn** [comment]
This is great! Are you collaborating with https://github.com/NVIDIA/bionemo-framework/pull/209? Would be good to synchronize efforts here and have that dataset make use of this once ready.

File: sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py (hunk 1 above)
Comment (skothenhill-nv) [reply sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py:29]: definitely needs tests, all of it haha.

**Comment by skothenhill-nv** [comment]
> This is great! Are you collaborating with #209? Would be good to synchronize efforts here and have that dataset make use of this once ready.

Yes, this is for that collaboration specifically. I shared the PR with Cristian last week.

**Pull request merged**